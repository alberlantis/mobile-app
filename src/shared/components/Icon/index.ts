export type { IconProps, IconName, IconType } from "./Icon";
export { default } from "./Icon";
