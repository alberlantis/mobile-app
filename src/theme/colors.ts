export default {
  BLACK: "#000",
  BLACK_INPUT: "#28292A",
  WHITE: "#FFF",
  WHITE_LIGHT: "rgba(255, 255, 255, 0.33)",
  ORANGE_PRIMARY_DARK: "#FF7849",
  ORANGE_PRIMARY_LIGHT: "#FF8D3B",
  BLUE_PRIMARY: "#5466D8",
  TRANSPARENT: "transparent",
};
