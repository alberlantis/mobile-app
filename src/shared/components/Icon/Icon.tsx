import React from "react";
import { Entypo, Feather } from "@expo/vector-icons";
import { ColorValue, StyleProp, TextStyle } from "react-native";

import colors from "src/theme/colors";

export type IconNames = {
  Entypo: keyof typeof Entypo.glyphMap;
  Feather: keyof typeof Feather.glyphMap;
};

interface IBaseIconProps {
  size?: number;
  color?: ColorValue;
  style?: StyleProp<TextStyle>;
}

export type IconProps<T extends keyof IconNames> = IBaseIconProps & {
  type: T;
  name: IconNames[T];
};

const Icon = <T extends keyof IconNames>({
  type,
  name,
  size = 24,
  color = colors.BLACK,
  style,
}: IconProps<T>) => {
  const iconMap = {
    Entypo,
    Feather,
  };
  const IconComponent = iconMap[type] as React.ComponentType<any>;
  return <IconComponent name={name} size={size} color={color} style={style} />;
};

export default Icon;
