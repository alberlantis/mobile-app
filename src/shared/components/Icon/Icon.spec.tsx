import { Entypo, Feather } from "@expo/vector-icons";
import { render, screen } from "@testing-library/react-native";

import colors from "src/theme/colors";
import Icon, { IconProps, IconNames } from "./Icon";

jest.mock("@expo/vector-icons", () => {
  const { Text } =
    jest.requireActual<typeof import("react-native")>("react-native");
  const Entypo = function (props: IconProps<"Entypo">) {
    return (
      <Text
        style={{ fontSize: props.size, color: props.color }}
      >{`Entypo-icon-${props.name}`}</Text>
    );
  };
  Entypo.glyphMap = {
    entypo1: "entypo1",
    entypo2: "entypo2",
    entypo3: "entypo3",
    entypo4: "entypo4",
  };
  const Feather = function (props: IconProps<"Feather">) {
    return (
      <Text
        style={{ fontSize: props.size, color: props.color }}
      >{`Feather-icon-${props.name}`}</Text>
    );
  };
  Feather.glyphMap = {
    feather1: "feather1",
    feather2: "feather2",
    feather3: "feather3",
    feather4: "feather4",
  };
  return {
    Entypo,
    Feather,
  };
});

const renderComponent = <T extends keyof IconNames>(props: IconProps<T>) => {
  render(<Icon {...props} />);
};

describe("Icon", () => {
  describe("when Entypo Icons", () => {
    it("should render all Entypo icons with default color and size", () => {
      const iconNames = Object.keys(
        Entypo.glyphMap,
      ) as (keyof typeof Entypo.glyphMap)[];
      iconNames.forEach((iconName) => {
        renderComponent<"Entypo">({ type: "Entypo", name: iconName });
        const icon = screen.getByText(`Entypo-icon-${iconName}`);
        expect(screen.toJSON()).toMatchSnapshot();
        expect(icon).toBeOnTheScreen();
        expect(icon).toHaveStyle({
          fontSize: 24,
          color: colors.BLACK,
        });
      });
    });

    it("should render all Entypo icons with given color and size", () => {
      const iconNames = Object.keys(
        Entypo.glyphMap,
      ) as (keyof typeof Entypo.glyphMap)[];
      iconNames.forEach((iconName) => {
        renderComponent<"Entypo">({
          type: "Entypo",
          name: iconName,
          size: 30,
          color: colors.ORANGE_PRIMARY_DARK,
        });
        const icon = screen.getByText(`Entypo-icon-${iconName}`);
        expect(screen.toJSON()).toMatchSnapshot();
        expect(icon).toBeOnTheScreen();
        expect(icon).toHaveStyle({
          fontSize: 30,
          color: colors.ORANGE_PRIMARY_DARK,
        });
      });
    });
  });

  describe("when Feather Icons", () => {
    it("should render all Feather icons with default color and size", () => {
      const iconNames = Object.keys(
        Feather.glyphMap,
      ) as (keyof typeof Feather.glyphMap)[];
      iconNames.forEach((iconName) => {
        renderComponent<"Feather">({ type: "Feather", name: iconName });
        const icon = screen.getByText(`Feather-icon-${iconName}`);
        expect(screen.toJSON()).toMatchSnapshot();
        expect(icon).toBeOnTheScreen();
        expect(icon).toHaveStyle({
          fontSize: 24,
          color: colors.BLACK,
        });
      });
    });

    it("should render all Feather icons with given color and size", () => {
      const iconNames = Object.keys(
        Feather.glyphMap,
      ) as (keyof typeof Feather.glyphMap)[];
      iconNames.forEach((iconName) => {
        renderComponent<"Feather">({
          type: "Feather",
          name: iconName,
          size: 30,
          color: colors.ORANGE_PRIMARY_DARK,
        });
        const icon = screen.getByText(`Feather-icon-${iconName}`);
        expect(screen.toJSON()).toMatchSnapshot();
        expect(icon).toBeOnTheScreen();
        expect(icon).toHaveStyle({
          fontSize: 30,
          color: colors.ORANGE_PRIMARY_DARK,
        });
      });
    });
  });
});
