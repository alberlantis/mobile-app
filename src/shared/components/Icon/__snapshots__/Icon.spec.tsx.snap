// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon when Entypo Icons should render all Entypo icons with default color and size 1`] = `
<Text
  style={
    {
      "color": "#000",
      "fontSize": 24,
    }
  }
>
  Entypo-icon-entypo1
</Text>
`;

exports[`Icon when Entypo Icons should render all Entypo icons with default color and size 2`] = `
<Text
  style={
    {
      "color": "#000",
      "fontSize": 24,
    }
  }
>
  Entypo-icon-entypo2
</Text>
`;

exports[`Icon when Entypo Icons should render all Entypo icons with default color and size 3`] = `
<Text
  style={
    {
      "color": "#000",
      "fontSize": 24,
    }
  }
>
  Entypo-icon-entypo3
</Text>
`;

exports[`Icon when Entypo Icons should render all Entypo icons with default color and size 4`] = `
<Text
  style={
    {
      "color": "#000",
      "fontSize": 24,
    }
  }
>
  Entypo-icon-entypo4
</Text>
`;

exports[`Icon when Entypo Icons should render all Entypo icons with given color and size 1`] = `
<Text
  style={
    {
      "color": "#FF7849",
      "fontSize": 30,
    }
  }
>
  Entypo-icon-entypo1
</Text>
`;

exports[`Icon when Entypo Icons should render all Entypo icons with given color and size 2`] = `
<Text
  style={
    {
      "color": "#FF7849",
      "fontSize": 30,
    }
  }
>
  Entypo-icon-entypo2
</Text>
`;

exports[`Icon when Entypo Icons should render all Entypo icons with given color and size 3`] = `
<Text
  style={
    {
      "color": "#FF7849",
      "fontSize": 30,
    }
  }
>
  Entypo-icon-entypo3
</Text>
`;

exports[`Icon when Entypo Icons should render all Entypo icons with given color and size 4`] = `
<Text
  style={
    {
      "color": "#FF7849",
      "fontSize": 30,
    }
  }
>
  Entypo-icon-entypo4
</Text>
`;

exports[`Icon when Feather Icons should render all Feather icons with default color and size 1`] = `
<Text
  style={
    {
      "color": "#000",
      "fontSize": 24,
    }
  }
>
  Feather-icon-feather1
</Text>
`;

exports[`Icon when Feather Icons should render all Feather icons with default color and size 2`] = `
<Text
  style={
    {
      "color": "#000",
      "fontSize": 24,
    }
  }
>
  Feather-icon-feather2
</Text>
`;

exports[`Icon when Feather Icons should render all Feather icons with default color and size 3`] = `
<Text
  style={
    {
      "color": "#000",
      "fontSize": 24,
    }
  }
>
  Feather-icon-feather3
</Text>
`;

exports[`Icon when Feather Icons should render all Feather icons with default color and size 4`] = `
<Text
  style={
    {
      "color": "#000",
      "fontSize": 24,
    }
  }
>
  Feather-icon-feather4
</Text>
`;

exports[`Icon when Feather Icons should render all Feather icons with given color and size 1`] = `
<Text
  style={
    {
      "color": "#FF7849",
      "fontSize": 30,
    }
  }
>
  Feather-icon-feather1
</Text>
`;

exports[`Icon when Feather Icons should render all Feather icons with given color and size 2`] = `
<Text
  style={
    {
      "color": "#FF7849",
      "fontSize": 30,
    }
  }
>
  Feather-icon-feather2
</Text>
`;

exports[`Icon when Feather Icons should render all Feather icons with given color and size 3`] = `
<Text
  style={
    {
      "color": "#FF7849",
      "fontSize": 30,
    }
  }
>
  Feather-icon-feather3
</Text>
`;

exports[`Icon when Feather Icons should render all Feather icons with given color and size 4`] = `
<Text
  style={
    {
      "color": "#FF7849",
      "fontSize": 30,
    }
  }
>
  Feather-icon-feather4
</Text>
`;
